yt_dlp/extractor/msn.py

```diff
@@ -1,167 +1,208 @@
-                    formats.extend(self._extract_m3u8_formats(
-                        format_url, display_id, 'mp4',
-                        m3u8_id='hls', fatal=False))
-                elif 'format=mpd-time-csf' in format_url:
-                    formats.extend(self._extract_mpd_formats(
-                        format_url, display_id, 'dash', fatal=False))
-                elif '.ism' in format_url:
-                    if format_url.endswith('.ism'):
-                        format_url += '/manifest'
-                    formats.extend(self._extract_ism_formats(
-                        format_url, display_id, 'mss', fatal=False))
-                else:
-                    format_id = file_.get('formatCode')
-                    formats.append({
-                        'url': format_url,
-                        'ext': 'mp4',
-                        'format_id': format_id,
-                        'width': int_or_none(file_.get('width')),
-                        'height': int_or_none(file_.get('height')),
-                        'vbr': int_or_none(self._search_regex(r'_(\d+)\.mp4', format_url, 'vbr', default=None)),
-                        'quality': 1 if format_id == '1001' else None,
-                    })
-
             subtitles = {}
-            for file_ in video.get('files', []):
-                format_url = file_.get('url')
-                format_code = file_.get('formatCode')
-                if not format_url or not format_code:
-                    continue
-                if str(format_code) == '3100':
-                    subtitles.setdefault(file_.get('culture', 'en'), []).append({
-                        'ext': determine_ext(format_url, 'ttml'),
-                        'url': format_url,
-                    })
+            for file in traverse_obj(json_data, ('videoMetadata', 'externalVideoFiles', lambda _, v: url_or_none(v['url']))):
+                url = file['url']
+                ext = determine_ext(url)
+                if ext == 'm3u8':
+                    fmts, subs = self._extract_m3u8_formats_and_subtitles(
+                        url, page_id, 'mp4', m3u8_id='hls', fatal=False)
+                    formats.extend(fmts)
+                    self._merge_subtitles(subs, target=subtitles)
+                elif ext == 'mpd':
+                    fmts, subs = self._extract_mpd_formats_and_subtitles(
+                        url, page_id, mpd_id='dash', fatal=False)
+                    formats.extend(fmts)
+                    self._merge_subtitles(subs, target=subtitles)
+                else:
+                    formats.append(
+                        traverse_obj(file, {
+                            'url': 'url',
+                            'format_id': ('format', {str}),
+                            'filesize': ('fileSize', {int}),
+                            'height': ('height', {int}),
+                            'width': ('width', {int}),
```

Comment:
I would use `int_or_none` for these
```suggestion
                            'filesize': ('fileSize', {int_or_none}),
                            'height': ('height', {int_or_none}),
                            'width': ('width', {int_or_none}),
```